### PR TITLE
[SPARK-29489][ML][PySpark] ml.evaluation support log-loss

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
`ml.MulticlassClassificationEvaluator` & `mllib.MulticlassMetrics` support log-loss

### Why are the changes needed?
log-loss is an important classification metric and is widely used in practice

### Does this PR introduce any user-facing change?
Yes, add new option ("logloss") and a related param `eps`

### How was this patch tested?
added testsuites & local tests refering to sklearn

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 1]
```diff
@@ -28,28 +28,30 @@ import org.apache.spark.sql.{DataFrame, Row}
...
+        case 2 => (r.getDouble(0), r.getDouble(1), 1.0, null)
+        case 3 => (r.getDouble(0), r.getDouble(1), r.getDouble(2), null)
+        case 4 => (r.getDouble(0), r.getDouble(1), r.getDouble(2), r.getSeq[Double](3).toArray)
+        case _ => throw new IllegalArgumentException(s"Expected Row of tuples, got $r")
+      }
     })
 
-
-  private val confusions = predictionAndLabels.map {
+  private lazy val confusions = predictionAndLabels.map {
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: If the metricName==logloss, then the confusion matrix is not needed, so I make this computation lazy.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 2]
```diff
@@ -28,28 +28,30 @@ import org.apache.spark.sql.{DataFrame, Row}
...
    */
   private[mllib] def this(predictionAndLabels: DataFrame) =
-    this(predictionAndLabels.rdd.map {
-      case Row(prediction: Double, label: Double, weight: Double) =>
-        (prediction, label, weight)
-      case Row(prediction: Double, label: Double) =>
-        (prediction, label, 1.0)
-      case other =>
-        throw new IllegalArgumentException(s"Expected Row of tuples, got $other")
+    this(predictionAndLabels.rdd.map { r =>
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: matching will not work in pyspark, so I have to use `r.get` instead.
`MultilabelMetrics` also deals with dataframe in this way.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 3]
```diff
@@ -237,4 +239,38 @@ class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[_ <: Product])
...
+  /**
+   * Returns the logLoss, aka logistic loss or cross-entropy loss.
+   * @param eps LogLoss is undefined for p=0 or p=1, so probabilities are
+   *            clipped to max(eps, min(1 - eps, p)).
+   */
+  @Since("3.0.0")
+  def logLoss(eps: Double = 1e-15): Double = {
+    require(eps > 0 && eps < 0.5, s"eps must be in range (0, 0.5), but got $eps")
+    val loss1 = - math.log(eps)
+    val loss2 = - math.log(1 - eps)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: `- math.log1p(-eps)`? because eps is going to be very small

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 4]
```diff
@@ -237,4 +239,38 @@ class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[_ <: Product])
    */
   @Since("1.1.0")
   lazy val labels: Array[Double] = tpByClass.keys.toArray.sorted
+
+  /**
+   * Returns the logLoss, aka logistic loss or cross-entropy loss.
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: You could just use a `@return` tag
Also log-loss rather than logLoss

File: python/pyspark/ml/evaluation.py [hunk 5]
```diff
@@ -365,18 +375,25 @@ class MulticlassClassificationEvaluator(JavaEvaluator, HasLabelCol, HasPredictio
                  "The beta value used in weightedFMeasure|fMeasureByLabel."
                  " Must be > 0. The default value is 1.",
                  typeConverter=TypeConverters.toFloat)
+    eps = Param(Params._dummy(), "eps",
```
Comment (srowen) [inline python/pyspark/ml/evaluation.py]: Do we really need this? it's kind of arbitrary. Do other libs have an epsilon for computing log loss? if a classifier is returning 1 or 0 the log loss really is undefined.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 6]
```diff
@@ -23,33 +23,35 @@ import scala.collection.mutable
...
 import org.apache.spark.rdd.RDD
-import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.DataFrame
 
 /**
  * Evaluator for multiclass classification.
  *
- * @param predictionAndLabels an RDD of (prediction, label, weight) or
- *                         (prediction, label) tuples.
+ * @param predictionAndLabels an RDD of (prediction, label, weight, probability) or
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: This is my only concern, that you have to plumb through a probability just to support this, but it's not out of the question.

File: python/pyspark/ml/evaluation.py (hunk 5 above)
Comment (zhengruifeng) [reply python/pyspark/ml/evaluation.py]: SKLearn has this param (default value is 1e-15), I follow sklearn's design.
Other lib like xgboost also use an epsilon to compute [log-loss](https://github.com/dmlc/xgboost/blob/master/src/objective/regression_loss.h#L72), but it's not exposed to end user and its default value is 1e-16.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala (hunk 6 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: Since current input do not provide probability for metrics like logloss, so I have to add it here.

File: python/pyspark/ml/evaluation.py (hunk 5 above)
Comment (srowen) [reply python/pyspark/ml/evaluation.py]: OK. I guess that's good enough reason to expose it. It's so niche that I almost think it's not worth it, but ...

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 7]
```diff
@@ -23,33 +23,35 @@ import scala.collection.mutable
...
+ *                            (prediction, label, weight) or (prediction, label) tuples.
  */
 @Since("1.1.0")
 class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[_ <: Product]) {
 
   /**
    * An auxiliary constructor taking a DataFrame.
-   * @param predictionAndLabels a DataFrame with two double columns: prediction and label
+   * @param predictionAndLabels a DataFrame with columns: prediction, label, weight(optional)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: Nit: spaces before paren

**Review by srowen — COMMENTED** [review]
Looking OK pending tests and one very minor comment

**Comment by zhengruifeng** [comment]
merged to master, thanks @srowen for reviewing!
